utils/ethabi.go

```diff
@@ -0,0 +1,265 @@
+package utils
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+	"strconv"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/tidwall/gjson"
+)
+
+const (
+	// FormatBytes encodes the output as bytes
+	FormatBytes = "bytes"
+	// FormatUint256 encodes the output as bytes containing a uint256
+	FormatUint256 = "uint256"
+	// FormatInt256 encodes the output as bytes containing an int256
+	FormatInt256 = "int256"
+	// FormatBool encodes the output as bytes containing a bool
+	FormatBool = "bool"
+)
+
+// ConcatBytes appends a bunch of byte arrays into a single byte array
+func ConcatBytes(bufs ...[]byte) ([]byte, error) {
+	buffer := bytes.NewBuffer([]byte{})
+	for _, b := range bufs {
+		_, err := buffer.Write(b)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buffer.Bytes(), nil
+}
+
+// EVMTranscodeBytes converts a json input to an EVM bytes array
+func EVMTranscodeBytes(value gjson.Result) ([]byte, error) {
+	switch value.Type {
+	case gjson.String:
+		return EVMEncodeBytes([]byte(value.Str))
+
+	case gjson.False:
+		return EVMEncodeBytes(EVMWordUint64(0))
+
+	case gjson.True:
+		return EVMEncodeBytes(EVMWordUint64(1))
+
+	case gjson.Number:
+		word, err := EVMWordSignedBigInt(big.NewInt(int64(value.Num)))
+		if err != nil {
+			return []byte{}, nil
+		}
+
+		return EVMEncodeBytes(word)
+
+	default:
+		return []byte{}, fmt.Errorf("unsupported encoding for value: %s", value.Type)
+	}
+}
+
+func roundToEVMWordBorder(length int) int {
+	mod := length % EVMWordByteLen
+	if mod == 0 {
+		return 0
+	}
+	return EVMWordByteLen - mod
+}
+
+// EVMEncodeBytes encodes arbitrary bytes as bytes expected by the EVM
+func EVMEncodeBytes(input []byte) ([]byte, error) {
+	length := len(input)
+	return ConcatBytes(
+		EVMWordUint64(uint64(length)),
+		input,
+		make([]byte, roundToEVMWordBorder(length)))
+}
+
+// EVMTranscodeBool converts a json input to an EVM bool
+func EVMTranscodeBool(value gjson.Result) ([]byte, error) {
+	var output uint64
+
+	switch value.Type {
+	case gjson.Number:
+		if value.Num != 0 {
+			output = 1
+		}
+
+	case gjson.String:
+		if len(value.Str) > 0 {
+			output = 1
+		}
+
+	case gjson.True:
+		output = 1
+
+	case gjson.JSON:
+		value.ForEach(func(key, value gjson.Result) bool {
+			output = 1
+			return false
+		})
+
+	case gjson.False, gjson.Null:
+
+	default:
+		panic(fmt.Errorf("unreachable/unsupported encoding for value: %s", value.Type))
+	}
+
+	return EVMWordUint64(output), nil
+}
+
+func parseNumericString(input string) (*big.Int, error) {
+	if HasHexPrefix(input) {
+		output, ok := big.NewInt(0).SetString(RemoveHexPrefix(input), 16)
+		if !ok {
+			return nil, fmt.Errorf("error parsing %s", input)
+		}
+		return output, nil
+	}
+
+	output, ok := big.NewInt(0).SetString(input, 10)
+	if !ok {
+		parseValue, err := strconv.ParseFloat(input, 64)
```

Comment:
@se3000 note that this will also lose precision if the string is encoded with a decimal or exponent encoding.